Action not found on dashboard on serdes v2

### Describe the bug

A very weird issue where serdesv2 doesn't find the action in a dashboard

### To Reproduce

Follow this exact steps at this exact order, please don't deviate a single step since this one is extremely difficult to reproduce:
1) set up origin and destination, from scratch
2) add a database (e.g. sample postgres db)
3) enable model actions on that database
4) do x-ray on people, save it
5) export on origin and import on destination (new commands, not dump and load)
6) now go to new->question->raw data-> people. Save the question. Turn into a model
7) go to model actions, create basic actions
8) add a new action-> "insert into people (1)" save it with any name
9) add a dashboard, in the dashboard add the model and a button that maps to the newly created action
10) Export and import, you should see the error


### Expected behavior

_No response_

### Logs

```
2023-07-22 00:39:23,982 INFO v2.load :: Loading Card fiCqhCw6mSXfFiqYHvYiI
clojure.lang.ExceptionInfo: Failed to load into database for [#ordered/map ([:model "Dashboard"] [:id "gmHEIwiD1i5Ffd4GSKiTw"])] {:path [#ordered/map ([:model "Dashboard"] [:id "gmHEIwiD1i5Ffd4GSKiTw"])], :deps-chain #{}}
	at metabase_enterprise.serialization.v2.load$load_one.invokeStatic(load.clj:70)
	at metabase_enterprise.serialization.v2.load$load_one.invoke(load.clj:33)
	at clojure.core.protocols$fn__8249.invokeStatic(protocols.clj:168)
	at clojure.core.protocols$fn__8249.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6886)
	at clojure.core$reduce.invoke(core.clj:6868)
	at metabase_enterprise.serialization.v2.load$load_metabase.invokeStatic(load.clj:95)
	at metabase_enterprise.serialization.v2.load$load_metabase.doInvoke(load.clj:83)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase_enterprise.serialization.cmd$fn__115199$v2_load__115200.invoke(cmd.clj:88)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at malli.core$_instrument$fn__13257.doInvoke(core.cljc:2526)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.cmd$call_enterprise.invokeStatic(cmd.clj:53)
	at metabase.cmd$call_enterprise.doInvoke(cmd.clj:43)
	at clojure.lang.RestFn.invoke(RestFn.java:439)
	at metabase.cmd$import.invokeStatic(cmd.clj:191)
	at metabase.cmd$import.doInvoke(cmd.clj:187)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.cmd$run_cmd$fn__117338.invoke(cmd.clj:299)
	at metabase.cmd$run_cmd.invokeStatic(cmd.clj:298)
	at metabase.cmd$run_cmd.invoke(cmd.clj:288)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.core$run_cmd.invokeStatic(core.clj:181)
	at metabase.core$run_cmd.invoke(core.clj:179)
	at metabase.core$_main.invokeStatic(core.clj:203)
	at metabase.core$_main.doInvoke(core.clj:198)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at metabase.bootstrap.main(Unknown Source)
Caused by: clojure.lang.ExceptionInfo: Could not find foreign key target - bad serdes-dependencies or other serialization error {:entity_id "lai4zes4g_UvWvzdBg5dS", :model "Action"}
	at metabase.models.serialization$_STAR_import_fk_STAR_.invokeStatic(serialization.clj:477)
	at metabase.models.serialization$_STAR_import_fk_STAR_.invoke(serialization.clj:456)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$memoize$fn__6946.doInvoke(core.clj:6388)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at clojure.core$update.invokeStatic(core.clj:6233)
	at clojure.core$update.invoke(core.clj:6223)
	at metabase.models.dashboard_card$fn__73407.invokeStatic(dashboard_card.clj:372)
	at metabase.models.dashboard_card$fn__73407.invoke(dashboard_card.clj:367)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.models.serialization$fn__45695.invokeStatic(serialization.clj:349)
	at metabase.models.serialization$fn__45695.invoke(serialization.clj:347)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.models.dashboard$fn__78233.invokeStatic(dashboard.clj:579)
	at metabase.models.dashboard$fn__78233.invoke(dashboard.clj:573)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase_enterprise.serialization.v2.load$load_one.invokeStatic(load.clj:67)
	... 49 more
Command failed with exception: Failed to load into database for [#ordered/map ([:model "Dashboard"] [:id "gmHEIwiD1i5Ffd4GSKiTw"])]
```

### Information about your Metabase installation

```JSON
v47.0-RC2
```


### Severity

P2ish

### Additional context

very difficult to consistently reproduce, but I managed to make it break 2 times in a row with those steps

--- a/enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj
@@ -6,7 +6,8 @@
    [metabase-enterprise.serialization.test-util :as ts]
    [metabase-enterprise.serialization.v2.extract :as extract]
    [metabase.models
-    :refer [Card
+    :refer [Action
+            Card
             Collection
             Dashboard
             DashboardCard
@@ -146,6 +147,17 @@
                                                                                        :filter [:>= [:field field-id nil] 18]
                                                                                        :aggregation [[:count]]}
                                                                                :database db-id}}]
+                       Card       [{model-id  :id
+                                    model-eid :entity_id}     {:name          "Some Model"
+                                                               :database_id   db-id
+                                                               :table_id      no-schema-id
+                                                               :collection_id coll-id
+                                                               :creator_id    mark-id
+                                                               :dataset       true
+                                                               :dataset_query {:query {:source-table no-schema-id
+                                                                                       :filter [:>= [:field field-id nil] 18]
+                                                                                       :aggregation [[:count]]}
+                                                                               :database db-id}}]
                        Card       [{c2-id  :id
                                     c2-eid :entity_id}        {:name          "Second Question"
                                                                :database_id   db-id
@@ -206,6 +218,11 @@
                                                                         :aggregation [[:count]]}
                                                                 :database db-id}}]
 
+                       Action     [{action-id    :id
+                                    action-eid   :entity_id}  {:name "Some action"
+                                                               :type :query
+                                                               :model_id model-id}]
+
                        Dashboard  [{dash-id  :id
                                     dash-eid :entity_id}      {:name          "Shared Dashboard"
                                                                :collection_id coll-id
@@ -254,7 +271,9 @@
                                                                   :fieldRef [:field "Average order total" {:base-type :type/Float}]
                                                                   :enabled true}]
                                                                 :column_settings
-                                                                {(str "[\"ref\",[\"field\"," field2-id ",null]]") {:column_title "Locus"}}}}]]
+                                                                {(str "[\"ref\",[\"field\"," field2-id ",null]]") {:column_title "Locus"}}}}]
+                       DashboardCard [_                       {:action_id action-id
+                                                               :dashboard_id other-dash-id}]]
       (testing "table and database are extracted as [db schema table] triples"
         (let [ser (serdes/extract-one "Card" {} (t2/select-one 'Card :id c1-id))]
           (is (schema= {:serdes/meta                 (s/eq [{:model "Card" :id c1-eid :label "some_question"}])
@@ -383,35 +402,35 @@
 
       (testing "Dashboards include their Dashcards"
         (let [ser (serdes/extract-one "Dashboard" {} (t2/select-one 'Dashboard :id other-dash-id))]
-          (is (schema= {:serdes/meta            (s/eq [{:model "Dashboard" :id other-dash :label "dave_s_dash"}])
-                        :entity_id              (s/eq other-dash)
-                        :ordered_cards
-                        [{:visualization_settings (s/eq {:table.pivot_column "SOURCE"
-                                                         :table.cell_column "sum"
-                                                         :table.columns
-                                                         [{:name "SOME_FIELD"
-                                                           :fieldRef [:field ["My Database" nil "Schemaless Table" "Some Field"] nil]
-                                                           :enabled true}
-                                                          {:name "sum"
-                                                           :fieldRef [:field "sum" {:base-type :type/Float}]
-                                                           :enabled true}
-                                                          {:name "count"
-                                                           :fieldRef [:field "count" {:base-type :type/BigInteger}]
-                                                           :enabled true}
-                                                          {:name "Average order total"
-                                                           :fieldRef [:field "Average order total" {:base-type :type/Float}]
-                                                           :enabled true}]
-                                                         :column_settings
-                                                         {"[\"ref\",[\"field\",[\"My Database\",\"PUBLIC\",\"Schema'd Table\",\"Other Field\"],null]]" {:column_title "Locus"}}})
-                          :created_at             LocalDateTime
-                          s/Keyword               s/Any}]
-                        :created_at             LocalDateTime
-                        s/Keyword               s/Any}
-                       ser))
+          (is (=? {:serdes/meta            [{:model "Dashboard" :id other-dash :label "dave_s_dash"}]
+                   :entity_id              other-dash
+                   :ordered_cards
+                   [{:visualization_settings {:table.pivot_column "SOURCE"
+                                              :table.cell_column "sum"
+                                              :table.columns
+                                              [{:name "SOME_FIELD"
+                                                :fieldRef [:field ["My Database" nil "Schemaless Table" "Some Field"] nil]
+                                                :enabled true}
+                                               {:name "sum"
+                                                :fieldRef [:field "sum" {:base-type :type/Float}]
+                                                :enabled true}
+                                               {:name "count"
+                                                :fieldRef [:field "count" {:base-type :type/BigInteger}]
+                                                :enabled true}
+                                               {:name "Average order total"
+                                                :fieldRef [:field "Average order total" {:base-type :type/Float}]
+                                                :enabled true}]
+                                              :column_settings
+                                              {"[\"ref\",[\"field\",[\"My Database\",\"PUBLIC\",\"Schema'd Table\",\"Other Field\"],null]]" {:column_title "Locus"}}}
+                     :created_at             LocalDateTime}
+                    {:action_id action-eid}]
+                   :created_at             LocalDateTime}
+                  ser))
           (is (not (contains? ser :id)))
 
-          (testing "and depend on all referenced cards, including those in visualization_settings"
+          (testing "and depend on all referenced cards and actions, including those in visualization_settings"
             (is (= #{[{:model "Card"       :id c2-eid}]
+                     [{:model "Action"     :id action-eid}]
                      [{:model "Database"   :id "My Database"}
                       {:model "Table"      :id "Schemaless Table"}
                       {:model "Field"      :id "Some Field"}]

--- a/enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj
@@ -147,8 +147,7 @@
                                                                                        :filter [:>= [:field field-id nil] 18]
                                                                                        :aggregation [[:count]]}
                                                                                :database db-id}}]
-                       Card       [{model-id  :id
-                                    model-eid :entity_id}     {:name          "Some Model"
+                       Card       [{model-id  :id}            {:name          "Some Model"
                                                                :database_id   db-id
                                                                :table_id      no-schema-id
                                                                :collection_id coll-id

--- a/src/metabase/models/dashboard.clj
+++ b/src/metabase/models/dashboard.clj
@@ -580,10 +580,11 @@
                         (t2/select-one 'DashboardCard :entity_id (:entity_id dashcard))))))
 
 (defn- serdes-deps-dashcard
-  [{:keys [card_id parameter_mappings visualization_settings]}]
+  [{:keys [action_id card_id parameter_mappings visualization_settings]}]
   (->> (mapcat serdes/mbql-deps parameter_mappings)
        (concat (serdes/visualization-settings-deps visualization_settings))
-       (concat (when card_id #{[{:model "Card" :id card_id}]}))
+       (concat (when card_id   #{[{:model "Card"   :id card_id}]}))
+       (concat (when action_id #{[{:model "Action" :id action_id}]}))
        set))
 
 (defmethod serdes/dependencies "Dashboard"

